File: docs-examples/example-java/src/test/java/io/micronaut/mqtt/docs/custom/type/ProductInfoTypeBinder.java
package io.micronaut.mqtt.docs.custom.type;

// tag::imports[]
import io.micronaut.context.annotation.Requires;
import io.micronaut.core.convert.ArgumentConversionContext;
import io.micronaut.core.convert.ConversionService;
import io.micronaut.core.type.Argument;
import io.micronaut.mqtt.bind.TypedMqttBinder;
import io.micronaut.mqtt.v5.bind.MqttV5BindingContext;
import org.eclipse.paho.mqttv5.common.packet.UserProperty;

import javax.inject.Singleton;
import java.util.*;
import java.util.stream.Collectors;
// end::imports[]

@Requires(property = "spec.name", value = "ProductInfoSpec")
// tag::clazz[]
@Singleton // <1>
public class ProductInfoTypeBinder implements TypedMqttBinder<MqttV5BindingContext, ProductInfo> { //<2>

    private final ConversionService<?> conversionService;

    ProductInfoTypeBinder(ConversionService<?> conversionService) { //<3>
        this.conversionService = conversionService;
    }

    @Override
    public Argument<ProductInfo> getArgumentType() {
        return Argument.of(ProductInfo.class);
    }

    @Override
    public void bindTo(MqttV5BindingContext context, ProductInfo value, Argument<ProductInfo> argument) {
        List<UserProperty> userPropertiesList = context.getProperties().getUserProperties();
        if (value.getSize() != null) {
            userPropertiesList.add(new UserProperty("productSize", value.getSize()));
        }
        userPropertiesList.add(new UserProperty("productCount", value.getCount().toString()));
        userPropertiesList.add(new UserProperty("productSealed", value.getSealed().toString()));
    }

    @Override
    public Optional<ProductInfo> bindFrom(MqttV5BindingContext context, ArgumentConversionContext<ProductInfo> conversionContext) {
        List<UserProperty> userPropertiesList = context.getProperties().getUserProperties();
        Map<String, String> userProperties = userPropertiesList.stream()
                .collect(Collectors.toMap(UserProperty::getKey, UserProperty::getValue));
        String size = userProperties.get("productSize");
        Optional<Long> count = Optional.ofNullable(userProperties.get("productCount"))
                .flatMap(value -> conversionService.convert(value, Long.class));
        Optional<Boolean> sealed = Optional.ofNullable(userProperties.get("productSealed"))
                .flatMap(value -> conversionService.convert(value, Boolean.class));

        if (count.isPresent() && sealed.isPresent()) {
            return Optional.of(new ProductInfo(size, count.get(), sealed.get()));
        } else {
            return Optional.empty();
        }
    }
}
// end::clazz[]

File: test-suite-python/src/test/python/micronaut/mqtt/docs/custom/type/ProductInfoTypeBinder.py
# tag::imports[]
from jakarta.inject import Singleton
from micronaut.core.convert import ArgumentConversionContext, ConversionService
from micronaut.core.type import Argument
from micronaut.mqtt.bind import TypedMqttBinder
from micronaut.mqtt.v5.bind import MqttV5BindingContext
from org.eclipse.paho.mqttv5.common.packet import UserProperty

from java.lang import Boolean, Long
from java.util import Optional

from .ProductInfo import ProductInfo
# end::imports[]

from micronaut.context.annotation import Requires


@Requires(property="spec.name", value="ProductInfoSpec")
# tag::clazz[]
@Singleton  # <1>
class ProductInfoTypeBinder(TypedMqttBinder):  # <2>

    def __init__(self, conversion_service: ConversionService):  # <3>
        self.conversion_service = conversion_service

    def getArgumentType(self) -> Argument:
        return Argument.of(ProductInfo)

    def bindTo(self, context: MqttV5BindingContext, value: ProductInfo, argument: Argument) -> None:
        user_properties_list = context.getProperties().getUserProperties()
        if value.size is not None:
            user_properties_list.add(UserProperty("productSize", value.size))
        user_properties_list.add(UserProperty("productCount", str(value.count)))  # <4>
        user_properties_list.add(UserProperty("productSealed", str(value.sealed).lower()))

    def bindFrom(self, context: MqttV5BindingContext, conversion_context: ArgumentConversionContext) -> Optional:
        user_properties = {p.getKey(): p.getValue() for p in context.getProperties().getUserProperties()}
        size = user_properties.get("productSize")
        count = Optional.ofNullable(user_properties.get("productCount")).flatMap(lambda value: self.conversion_service.convert(value, Long))
        sealed = Optional.ofNullable(user_properties.get("productSealed")).flatMap(lambda value: self.conversion_service.convert(value, Boolean))

        if count.isPresent() and sealed.isPresent():
            return Optional.of(ProductInfo(size, count.get(), sealed.get()))  # <5>
        else:
            return Optional.empty()
# end::clazz[]
